rfc/005-soft-chain-upgrades.md

```diff
@@ -0,0 +1,473 @@
+If one of these doesn't match we drop the block and peer and continue looking
+for another. (Remember light clients only have `NextValidatorHash` and have no
+record of app state)
+8. Go back to 1 and recur until we reach a point where the `DerivedBlockHash`
+is nil. This indicates the crossover from the migrated blocks to the original
+ones.
+9. When validating this block we return to the normal verification process.
+This means that instead of using the `DerivedBlockHash` we take the signatures
+in the commit and verify it against the hash of the entire header.
+This means that we not only trust the `NextValidatorHash` and `AppHash` but can
+trust the entire header. This includes LastBlockID.
+10. Verify that LastBlockID equates to the hash of the last migrated block. If
+this is equal then we have essentially verified the entire migration history and
+we can trust that the state transitions that have been applied are correct.
+11. Proceed as normal until the node can switch to consensus.
+
+Another way of viewing `DerivedBlockHash` is as the remnants of the hashes/data
+in the old header that we don't want to bring across because we have changed it.
+
+In terms of the actual upgrade, when consensus is reached on a block that causes
+the block version to increment, the next block proposed should not only have the
+new block structure but the LastBlockID should be that of the migrated block not
+the original one. This means that before the next block of the new version can
+be agreed upon everyone participating in consensus must have migrated the entire
+history in order to generate the "golden hash", the cumulative hash of the
+migrated history. If this is three years of blocks this migration could take a
+long time before consensus can continue hence it's probably an important aspect
+to attempt to make migration asynchronous so nodes already know this golden hash
+before the upgrade actually happens.
+
+This could be done by executing a migration script that works "offline" by
+iterating through the block store and making copies of the new header only so
+that nodes know what the migrated header hash would be and when consensus
+reached the upgrade height, could smoothly transition to the newly migrated
+blocks
+
+### Implications
+
+**Positive**
+
+- The network only needs to concern itself with the latest block version
+- Some migrations would be backwards compatible meaning we could restore the
+prior block version. This would allow external clients on prior block versions
+to still be able to process the blocks.
+- Async upgrading. Nodes choose when they want to upgrade the binary and
+Tendermint handles the actual transition.
+- In the future it may be possible to support application migrations. This is
+where the application can upgrade the Tx data structures. Generally, this option
+might offer a broader set of changes that can be soft upgraded.
+
+**Negative**
+
```

Comment:
I'd add to the negatives that this is also the more complex solution to implement in the short term.

Reply:
(This is sort of implied, but good to call out explicitly.)

Reply:
Also: Requires building some kind of separate migration tool, which probably wouldn't get exercised that often and would be at-risk for getting musty and outdated